packages/cursorless-engine/src/processTargets/modifiers/scopeHandlers/VisibleScopeHandler.ts

```diff
@@ -0,0 +1,35 @@
+import { TextEditor, Position } from "@cursorless/common";
+import { Direction, ScopeType } from "@cursorless/common";
+import { PlainTarget } from "../../targets";
+import BaseScopeHandler from "./BaseScopeHandler";
+import { TargetScope } from "./scope.types";
+
+export default class VisibleScopeHandler extends BaseScopeHandler {
+  public readonly scopeType = { type: "visible" } as const;
+  public readonly iterationScopeType = { type: "document" } as const;
+  protected readonly isHierarchical = false;
+
+  constructor(_scopeType: ScopeType, _languageId: string) {
+    super();
+  }
+
+  protected *generateScopeCandidates(
+    editor: TextEditor,
+    _position: Position,
+    _direction: Direction,
+  ): Iterable<TargetScope> {
+    for (const range of editor.visibleRanges) {
```

Comment:
This will interact with folding in an interesting way. If you say "take visible" from one side of a fold, you'll only get up until the fold. If you say "take next visible" you'll get the code between the next fold until the following. To refer to all sections between folds, you'd say "take every visible".

I'd be tempted to either

- yield a single scope whose range is the union all the visible ranges, or
- yield a single scope that returns all the visible regions as a single list in its `getTargets`. I'd consider making the domain of that scope just be the document.

Reply:
There's a third option, which is to reject `visible` as an allowed scope when folds are present. That sidesteps the question of what the correct behavior is, and punts the decision to the future, where we could loosen that constraint.

I lean towards your second option, and will implement that to play with, but am open to guidance. :)

Reply:
Implemented, played with second option. I'm happy with it. Another use case: fixing variable shadowing. Fold all shadowed inner scopes, work on every instance in what remains visible. (You can do that with take and give and this and that, but folding is more stateful, which can be useful.)

Reply:
Although that makes me wonder whether visible should be a modifier.

`from visible file take every instance air`
`from visible func fine take every instance air`

Nah, probably not.

Reply:
If you revert to your original approach, then you could say "from every visible funk" to target all visible ranges that touch the current function. Not necessarily suggesting you do that, but just fwiw

Reply:
That's true! But on balance, I think I prefer the simpler semantics. I'm happy with what is implemented now.

Reply:
Brain dump:

- This is the first example of a scope that's not purely a function of the document itself
- Being a scope type here doesn't buy us anything useful over just making it a modifier (eg we can't leverage "every visible", "next visible", etc even tho the grammar now allows them). However that's also true or document scope ("file")
- Having it be a mark would preclude ability to say "visible air" to target visible in another split, so we'd prob want it to at least be a modifier
- Having it be modifier with intersection as you proposed is interesting but might be unintuitive when you have a nonempty selection and say "take visible": that would then intersect your selection with visible
- it might actually be useful to say "next visible" / "first visible" etc. I think the latter does work with your impl fwiw
- There is an interesting parallel here to a hypothetical scope for referring to your selections, eg "first selection", or "selection air" to refer to the selection in the split containing air